### DISCUSSION TITLE
Memory leak in NIC?  NBP is too big to fit into free base memory

### DISCUSSION THREAD

**Discussion by BillyCroan** [ipxe/ipxe#275] [Q&A]
I have some hardware that will retry booting over and over if there are no usable boot devices.

The way I have iPXE set up, a machine will look for a specific script for its mac address, and if not found, then it waits for a keypress for a few seconds to display a menu of tools, and if it sees no keypress, then my script exits.

This mostly works as expected, but for one thing:  After 'a few minutes' of cycling through boot devices like this, my hardware inevitably starts displaying "NBP is too big to fit into free base memory", instead of trying to search for the script for its mac address.  If I power cycle the machine it goes back to normal for a few minutes.

I'd like to know that if a machine is "looping" looking for a boot device, and then I create that file for its mac address, that within a few minutes, it will pick it up and boot, but because of what I think is memory exhaustion in the NIC itself, it does not.

My first instinct is to cause a reboot at the end of my script instead of exiting.  But something seems ugly about that.  Have others run into this?  It can't be a bug in ipxe can it?  Is there a way to mitigate memory exhaustion in the nic firmware?

Is there a command I can use in an ipxe script to re-initialize the NIC, or clear its memory on exit of my script?

**Comment by NiKiZe** [comment]
To help with something like this, it is a loot easier if we can see the script.

Are your script chaining into iPXE over and over again, or is it only the script?
have you tried running imgstat to see if there is multiple scripts loaded, and have you tried to automatically unload scripts? see [chain autoofree and replace](https://ipxe.org/cmd/chain)

also which ipxe are you using? undionly.kpxe or snponly.efi etc (and yes every detail of that name could be important) or better yet, what is the command you are using to build iPXE ?

**Reply by mcb30** [reply]
@BillyCroan A copy of your script and a video of your cycling boot attempt would probably clarify precisely what is happening.

**Reply by BillyCroan** [reply]
Here's my [scripts](https://github.com/ipxe/ipxe/files/6079813/script.txt) and a [video](https://youtu.be/DF4q0Ylp_c8)
It doesn't take but 3-4 minutes before it starts complaining NBP too big.

**Reply by mcb30** [reply]
Thanks.  Your video doesn't show how you reached that point.  From examining your scripts `boot.ipxe` and `menu.ipxe`, I'm guessing that you are building `undionly.kpxe` with `EMBED=boot.ipxe`.

Your `menu.ipxe` includes the line:
```
choose --timeout ${menu-timeout} --default ${menu-default} selected || goto cancel
```
However, there is no `:cancel` label defined anywhere in the script.  On a timeout (or explicit cancellation) your script will therefore attempt to `goto` a nonexistent label.  This will fail and cause the script to exit immediately.

Assuming that you are using `undionly.kpxe`, then this will subsequently cause iPXE to exit by asking the BIOS to boot from the next device (via `INT 18`).  This is likely to leave part of the Intel PXE stack resident in base memory, since there is no clean way for it to be removed.

I would recommend that you redesign your retry mechanism so that control stays within the iPXE script, rather than exiting back to the BIOS on each iteration.  Exiting to the BIOS is notoriously unreliable and you are unlikely to be able to make it work consistently as part of any kind of retry loop.

**Comment by BillyCroan** [comment]
On Wed, Mar 3, 2021 at 6:48 PM Michael Brown <notifications@github.com>
wrote:

> Thanks. Your video doesn't show how you reached that point.
>
To show from the moment of boot to the moment the error starts, the video
needs to be several minutes long.  I didn't realise that you wanted all
that so I hadnt included it.  Here's a new video
<https://youtu.be/az8Ivvi69UE> that captures that time period. Sorry it's
not more zoomed in.

> From examining your scripts boot.ipxe and menu.ipxe, I'm guessing that
> you are building undionly.kpxe with EMBED=boot.ipxe.
>
I'm not sure about that.  I am unfamiliar with EMBED=.
I searched my git repo for it and all I found was:
/root/gits/ipxe/src/Makefile.housekeeping:EMBED         :=
$(EMBEDDED_IMAGE) # Maintain backwards compatibility

If you're wondering how/why my ipxe searches for boot.ipxe, I'll share this
snippet from my dhcpd conf:
  #If dhcp client is the ipxe client I compiled and made (versus the dhcp
client of a NIC rom, OS, or libvirt vm)
  if exists ipxe.https {
    #Then give it our ipxe entry config
    filename "http://InstallNode/boot.ipxe";
  } else {
    #Else, give it a tftp path to download a modern ipxe build
    next-server InstallNode.InstallNet.Corp.UnrealServers.net; #Full suffix
is needed because this is resolved to ip on InstallNode when dhcp starts
    filename "undionly.kpxe";
  }

#  option ipxe.no-pxedhcp 1;
  option routers 192.168.13.1;
  option domain-name "InstallNet.Corp.UnrealServers.net";
  option domain-name-servers DNS.InstallNet.Corp.UnrealServers.net;
  default-lease-time 600;
  max-lease-time 7200;
  option log-servers 192.168.13.1;

So, most NICs  get filename and nextserver to cause them to fetch
undionly.kpxe over tftp.  Then.  When MY compiled ipxe is running on the
nic, it dhcp's again, and  gets the http url of boot.ipxe

> Your menu.ipxe includes the line:
>
> choose --timeout ${menu-timeout} --default ${menu-default} selected || goto cancel
>
> However, there is no :cancel label defined anywhere in the script.
>
Good eye!  I've corrected that.  It now reads || exit

> On a timeout (or explicit cancellation) your script will therefore attempt
> to goto a nonexistent label. This will fail and cause the script to exit
> immediately.
>
I wanted it to exit immediately.  But I should have coded it thusly of
course,  thanks for spotting it.

> Assuming that you are using undionly.kpxe, then this will subsequently
> cause iPXE to exit by asking the BIOS to boot from the next device (via INT
> 18). This is likely to leave part of the Intel PXE stack resident in base
> memory, since there is no clean way for it to be removed.
>
THAT.  Why is that?  That feels like the jackpot to me.  Physically
rebooting the node seems to remove it.  Why doesn't it clear when the
system cycles to trying the next boot device.

> I would recommend that you redesign your retry mechanism so that control
> stays within the iPXE script, rather than exiting back to the BIOS on each
> iteration.
>
I have many different types of hardware, each with a different set of
possible boot devices.  I depend on each system's bios to know about all of
the types that could be valid on a given system.  Is it possible for ipxe
to enumerate the boot device list from the bios and iterate over that list
one by one?

> Exiting to the BIOS is notoriously unreliable
>
Why is that?

**Reply by mcb30** [reply]
> ```
>   #If dhcp client is the ipxe client I compiled and made
>   if exists ipxe.https {
>     #Then give it our ipxe entry config
>     filename "http://InstallNode/boot.ipxe";
>   } else {
>     #Else, give it a tftp path to download a modern ipxe build
>     next-server InstallNode.InstallNet.Corp.UnrealServers.net; 
>     filename "undionly.kpxe";
>   }
> ```

Thanks: this `dhcpd.conf` fragment clarifies what you're doing.

> > I would recommend that you redesign your retry mechanism so that control stays within the iPXE script, rather than exiting back to the BIOS on each iteration.
> I have many different types of hardware, each with a different set of possible boot devices. I depend on each system's bios to know about all of the types that could be valid on a given system. Is it possible for ipxe to enumerate the boot device list from the bios and iterate over that list one by one?

So, my question is: once the system has been able to successfully get as far as running an iPXE script using your preferred build of iPXE, why do you ever need to exit back to try booting from another NIC?

(Conversely: if you have a NIC that is able to get link-up, DHCP, run iPXE, and execute your script, then what circumstances would make this not be the "correct" NIC from which to boot?)

> > Assuming that you are using undionly.kpxe, then this will subsequently cause iPXE to exit by asking the BIOS to boot from the next device (via INT 18). This is likely to leave part of the Intel PXE stack resident in base memory, since there is no clean way for it to be removed.
>
> THAT. Why is that? That feels like the jackpot to me. Physically rebooting the node seems to remove it. Why doesn't it clear when the system cycles to trying the next boot device.
>
> > Exiting to the BIOS is notoriously unreliable
>
> Why is that?

Rebooting the system will clear everything, since the whole memory contents get reset to (effectively) a power-on state.

When using `undionly.kpxe` (or `ipxe.pxe`, or any `*.pxe` iPXE binary), there is no available return path to the BIOS.  The portion of the vendor PXE code that would comprise the return path gets unloaded during iPXE initialisation.  If iPXE were to attempt to return (via the normal `lret` instruction) then it would end up executing random code (and would most probably crash the machine).

Once the vendor PXE stack has been unloaded, the closest available approximation of "returning" is to use `INT 18` to request that the BIOS boot the next device.  Unfortunately this is fairly undefined behaviour: there is nothing in any BIOS specification that dictates how the BIOS is supposed to behave when this happens.  For example: many HP BIOSes will treat an `INT 18` after an attempted network boot as meaning "there was probably a transient network error, best to retry booting from that exact same device again".

There are two ways you can avoid this undefined behaviour:

1. (strongly recommended): Keep the flow of control within your iPXE script once you have reached it, rather than exiting all the way back to the BIOS
2. (not recommended, may work on some systems, will definitely fail on others): Use the `undionly.kkpxe` binary instead of `undionly.kpxe`.  This binary skips unloading the vendor PXE code and so **will** have a defined return path at the point of exit.  However, the fact that the vendor PXE code remains loaded and active while iPXE is also running (and controlling the same NIC) may lead to other, more subtle, undefined behaviour.

**Comment by BillyCroan** [comment]
On Thu, Mar 4, 2021 at 7:10 AM Michael Brown <notifications@github.com>
wrote:

> once the system has been able to successfully get as far as running an
> iPXE script using your preferred build of iPXE, why do you ever need to
> exit back to try booting from another NIC?
>
From another NIC, I don't.  We've not ever needed to do that.

But very often, a node runs ipxe, does not find machine-specific
instruction, the user does not interact to request the netboot menu, and we
want the machine to continue to boot from hard disk or usb stick or raid
card or maybe on a very rare blue moon usb cdrom.  I imagine that I could
probably code my ipxe script to search for and try to boot those things.
But would they boot the same via ipxe as they would via the bios?

Would installer environments chainloaded through ipxe install the system
expecting that it always boot via ipxe (which won't be the case)?  Those
are the things I worry about when using ipxe to chain load a local
bootloader and not using the bios' original boot order system.  I think of
the old grub hd0 numbering and how it associates to sda,b,c etc. , which is
different if you boot to rescue media and chroot in, than it is in a live
working system.  Of course that's not what's going on here, but i worry
something similar could happen.

> Rebooting the system will clear everything, since the whole memory
> contents get reset to (effectively) a power-on state.
>
It's too bad there's not a command I can use in the ipxe script to do that
during exit.

> When using undionly.kpxe (or ipxe.pxe, or any *.pxe iPXE binary), there
> is no available return path to the BIOS.
>
Which binaries do have an available return path to the bios? (nvm, as I
read on later, the kk one does it.)

> The portion of the vendor PXE code that would comprise the return path *gets
> unloaded during iPXE initialisation*.
>
It sounds like that is the crux of it.  kkpxe sounds like it's worth a try.

>  For example: many HP BIOSes will treat an INT 18 after an attempted
> network boot as meaning "there was probably a transient network error, best
> to retry booting from that exact same device again".
>
Frustrating.  But come to think of it, I think I recall seeing that before
now...

> There are two ways you can avoid this undefined behaviour:
>
>    1. (strongly recommended): Keep the flow of control within your iPXE
>    script once you have reached it, rather than exiting all the way back to
>    the BIOS
>    2. (not recommended, may work on some systems, will definitely fail on
>    others): Use the undionly.kkpxe binary instead of undionly.kpxe. This
>    binary skips unloading the vendor PXE code and so *will* have a
>    defined return path at the point of exit. However, the fact that the vendor
>    PXE code remains loaded and active while iPXE is also running (and
>    controlling the same NIC) may lead to other, more subtle, undefined
>    behaviour.
>
> I didn't see a .kkpxe in there.  but I figured out it's pretty easy to
build, by just running 'make bin/undionly.kkpxe' in the src dir from
https://forum.ipxe.org/showthread.php?tid=1037

I went ahead and made kkkpxe as well which I might try if kk doesn't solve
it.

Does it matter what filename or extension I use on the tftp server and in
dhcp's filename arg?

I don't want to call it undionly.kpxe any more if I'm really using a kkpxe
file, and don't want to change it when/if I tweak this again.  I'm thinking
of just calling it nbp, and using a symlink named nbp to point to the type
and revision of bootloader I want to use.  (I've retained every compiled
version I've made for the last 4 years.)

I want to say I've ran into trouble before renaming netboot kernels
though.  if the extension changed then the bootloader tries to decompress
them when it can't or vice versa.

kkpxe has been holding steady fo 5+ minutes now while writing this.  And I
see the (vendor?) ipxe stack printing "exiting intel boot agent".
Excellent!  I  think this fixed it though I'll look out for other weirdness.

Thanks Michael!

**Reply by mcb30** [reply]
> From another NIC, I don't. We've not ever needed to do that. But very often, a node runs ipxe, does not find machine-specific instruction, the user does not interact to request the netboot menu, and we want the machine to continue to boot from hard disk or usb stick or raid card or maybe on a very rare blue moon usb cdrom. I imagine that I could probably code my ipxe script to search for and try to boot those things. But would they boot the same via ipxe as they would via the bios? Would installer environments chainloaded through ipxe install the system expecting that it always boot via ipxe (which won't be the case)? Those are the things I worry about when using ipxe to chain load a local bootloader and not using the bios' original boot order system. I think of the old grub hd0 numbering and how it associates to sda,b,c etc. , which is different if you boot to rescue media and chroot in, than it is in a live working system. Of course that's not what's going on here, but i worry something similar could happen.

That's fine: you can viably exit back to the BIOS once you have decided for sure that you will not be booting from the network.  There is a reasonable chance that this will work at least once on most platforms.  It's the looping that is likely to expose issues such as the one you have seen.

> Rebooting the system will clear everything, since the whole memory contents get reset to (effectively) a power-on state.
> It's too bad there's not a command I can use in the ipxe script to do that during exit.

The only command that can logically do this is https://ipxe.org/cmd/reboot

> Does it matter what filename or extension I use on the tftp server and in dhcp's filename arg? I don't want to call it undionly.kpxe any more if I'm really using a kkpxe file, and don't want to change it when/if I tweak this again.

No, the filename or extension doesn't matter to either vendor PXE ROMs or to iPXE.  Some other bootloaders may care about it.

> kkpxe has been holding steady fo 5+ minutes now while writing this. And I see the (vendor?) ipxe stack printing "exiting intel boot agent". Excellent! I think this fixed it though I'll look out for other weirdness.

Good luck!
